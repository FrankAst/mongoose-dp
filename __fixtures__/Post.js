// @flow
/* eslint-disable no-param-reassign, func-names */

import mongoose, { type MongooseSchema } from 'mongoose';
import DiffPlugin from '../src/index';
import DB from './db';

DB.init();

export const PostSchema: MongooseSchema<PostDoc> = new mongoose.Schema(
  {
    title: {
      type: String,
      description: 'Provider who send Post',
    },
    subjects: [
      {
        _id: { track_diff: false },
        name: { type: String, track_diff: false },
      },
    ],
    updatedAt: {
      type: Date,
      track_diff: false,
    },
  },
  {
    timestamps: true,
    collection: 'post',
  }
);

export class PostDoc /* :: extends Mongoose$Document */ {
  title: string;
  subjects: Array<{ name: string }>;

  // TODO: find out solution for flow to use `DiffModelT` instead of `any`
  /* :: static diffModel(): any {}  */
}

PostSchema.plugin(DiffPlugin);
PostSchema.loadClass(PostDoc);

export const Post = DB.data.model('Post', PostSchema);
