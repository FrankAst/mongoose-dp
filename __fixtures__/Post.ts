/* eslint-disable no-await-in-loop */
// @flow
/* eslint-disable no-param-reassign, func-names */

import mongoose, { Schema, Document, Model } from 'mongoose';
import DiffPlugin, { IDiffModel } from '../src/index';
import DB from './db';

DB.init();

export interface IPostDoc extends Document {
  title: string;
  subjects: Array<{ name: string }>;
}

interface IPostModel extends Model<IPostDoc> {
  diffModel(): IDiffModel;
  createDifferentSubjects(findObj: string, count: number): Promise<void>;
}

export const PostSchema: Schema<IPostDoc, IPostModel> = new mongoose.Schema(
  {
    title: {
      type: String,
      description: 'Provider who send Post',
    },
    subjects: [
      {
        _id: { type: mongoose.Schema.Types.ObjectId, track_diff: false },
        name: { type: String },
      },
    ],
    createdAt: {
      type: Date,
      track_diff: false,
    },
    updatedAt: {
      type: Date,
      track_diff: false,
    },
  },
  {
    timestamps: true,
    collection: 'post',
  }
);

// for test purposes
PostSchema.statics.createDifferentSubjects = async function (
  this: IPostModel,
  title: string,
  count: number
): Promise<void> {
  for (let i = 1; i <= count; i += 1) {
    const post = await this.findOne({ title }).exec();
    if (post) {
      post.subjects.push({ name: `name_${i}` });
      await post.save();
    }
  }
};

PostSchema.plugin(DiffPlugin<IPostDoc>());

export const Post = DB.data.model<IPostDoc, IPostModel>('Post', PostSchema);
