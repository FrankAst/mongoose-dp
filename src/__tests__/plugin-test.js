// @flow

import { Post, type PostDoc } from '../../__fixtures__/Post';

jest.mock('../../__fixtures__/db.js');
jasmine.DEFAULT_TIMEOUT_INTERVAL = 10000;

describe('mongoose-dp', () => {
  it('return diff model', async () => {
    const Diff = Post.diffModel();
    const diff = new Diff({ p: ['title'], c: [{ p: ['way'] }] });

    expect(typeof Diff).toBe('function');
    expect(typeof Diff.findByDocId).toBe('function');
    expect(typeof Diff.createDiff).toBe('function');
    expect(typeof Diff.findAfterVersion).toBe('function');
    expect(typeof Diff.findBeforeVersion).toBe('function');
    expect(typeof Diff.revertToVersion).toBe('function');
    // expect(typeof Diff.getNextVersion).toBe('function');
    expect(diff.c[0].p[0]).toBe('way');
  });

  it('save diffs', async () => {
    // need to return doc from db for invoking `init` hook,
    // because there is no sense to save initial doc in diffs
    await Post.create({ title: 'test', subjects: [{ name: 'matsdcsdch' }] });
    const post: PostDoc = (await Post.findOne({ title: 'test' }).exec(): any);
    post.title = 'updated';
    post.subjects = [{ name: 'math' }, { name: 'air' }];
    await post.save();

    const Diff = Post.diffModel();
    const diffs = await Diff.findByDocId(post._id);

    expect(Array.isArray(diffs)).toBeTruthy();
    expect(diffs[0].c).toMatchSnapshot();
    expect(diffs[0].v).toBe(1);
  });

  describe('save array diffs properly', () => {
    it('add new element', async () => {
      await Post.create({ title: 'newElement', subjects: [{ name: 'test' }] });
      const post: PostDoc = (await Post.findOne({
        title: 'newElement',
      }).exec(): any);
      post.subjects = [{ name: 'test' }, { name: 'new test' }];
      await post.save();

      const Diff = Post.diffModel();
      const diffs = await Diff.findByDocId(post._id);
      expect(diffs[0].c).toMatchInlineSnapshot(`
        CoreMongooseArray [
          Object {
            "i": 1,
            "it": Object {
              "k": "N",
              "r": Object {
                "name": "new test",
              },
            },
            "k": "A",
            "p": Array [
              "subjects",
            ],
          },
        ]
      `);
    });

    it('edit existed element', async () => {
      await Post.create({
        title: 'existedElement',
        subjects: [{ name: 'was' }],
      });
      const post: PostDoc = (await Post.findOne({
        title: 'existedElement',
      }).exec(): any);
      post.subjects = [{ name: 'become' }];
      await post.save();

      const Diff = Post.diffModel();
      const diffs = await Diff.findByDocId(post._id);
      expect(diffs[0].c).toMatchInlineSnapshot(`
        CoreMongooseArray [
          Object {
            "k": "E",
            "l": "was",
            "p": Array [
              "subjects",
              "0",
              "name",
            ],
            "r": "become",
          },
        ]
      `);
    });

    it('delete element', async () => {
      await Post.create({
        title: 'deleteElement',
        subjects: [{ name: 'one' }, { name: 'two' }, { name: 'three' }],
      });
      const post: PostDoc = (await Post.findOne({
        title: 'deleteElement',
      }).exec(): any);
      post.subjects = [{ name: 'one' }, { name: 'three' }];
      await post.save();

      const Diff = Post.diffModel();
      const diffs = await Diff.findByDocId(post._id);
      expect(diffs[0].c).toMatchInlineSnapshot(`
        CoreMongooseArray [
          Object {
            "i": 2,
            "it": Object {
              "k": "D",
              "l": Object {
                "name": "three",
              },
            },
            "k": "A",
            "p": Array [
              "subjects",
            ],
          },
          Object {
            "k": "E",
            "l": "two",
            "p": Array [
              "subjects",
              "1",
              "name",
            ],
            "r": "three",
          },
        ]
      `);
    });
  });
});
