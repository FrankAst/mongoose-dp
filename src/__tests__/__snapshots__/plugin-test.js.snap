// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mongoose-dp save diffs 1`] = `
CoreMongooseArray [
  Object {
    "k": "E",
    "l": "test",
    "p": Array [
      "title",
    ],
    "r": "updated",
  },
  Object {
    "i": 1,
    "it": Object {
      "k": "N",
      "r": Object {
        "name": "air",
      },
    },
    "k": "A",
    "p": Array [
      "subjects",
    ],
  },
  Object {
    "k": "E",
    "l": "matsdcsdch",
    "p": Array [
      "subjects",
      "0",
      "name",
    ],
    "r": "math",
  },
]
`;
